File: LeetCode75/88_Merge_Sorted_Array/88_Merge_Sorted_Array.ts
/**
 Do not return anything, modify nums1 in-place instead.
 */
function merge(nums1: number[], m: number, nums2: number[], n: number): void {
    // nums1 と nums2 を後ろから比較してマージ
    let p1 = m - 1; // nums1の有効部分の最後のインデックス
    let p2 = n - 1; // nums2の最後のインデックス
    let p = m + n - 1; // nums1の最後のインデックス（全体の長さ）

    // nums1とnums2の要素を後ろから比較して、大きい方を nums1 に格納
    while (p1 >= 0 && p2 >= 0) {
        if (nums1[p1] > nums2[p2]) {
            nums1[p] = nums1[p1];
            p1--;
        } else {
            nums1[p] = nums2[p2];
            p2--;
        }
        p--;
    }

    // nums2に要素が残っている場合、nums1の先頭部分にコピー
    while (p2 >= 0) {
        nums1[p] = nums2[p2];
        p2--;
        p--;
    }
}

File: LeetCode75/88_Merge_Sorted_Array/88_Merge_Sorted_Array.kt
function merge(num1: number[],m: number, num2: number[], n: number):void {
  if (m == 0) {
    console.log(num2);
    return;
  } else if(n == 0) {
    console.log(num1);
    return;
  }

  num1 = [...num1.slice(0, m), ...num2];
}

function main() {
  let num1 = [1,2,3];
  const m = 2;
  const num2 = [4,5,6];
  const n = 3;

  const result = merge(num1, m, num2, n);
  console.log(result);
};

main();
